### PR TITLE
Controller options and patch

### PR DESCRIPTION
**Pull request by sarbogast**
Added automatic handling of OPTIONS requests on ResourceController
And I also added a modify method to handle PATCH requests

### CODE REVIEW

**Comment by LoganWright** [comment]
@sarbogast thanks for continuing w/ some awesome contributions! I might be missing something, but it looks like the `modify` route might not be getting registered. Is it somewhere I can't see?

**Comment by sarbogast** [comment]
Yes, in Application+Route.swift:

self.patch(path, Resource.Item.self) { request, item in
            return try controllerFactory().modify(request, item:item)
        }

**Comment by LoganWright** [comment]
@sarbogast wow, don't know how I missed that, must need some more coffee or something! Thanks for pointing it out.

File: Sources/Development/Controllers/UserController.swift [hunk 1]
```diff
@@ -44,6 +44,13 @@ class UserController: Controller {
         return user.makeJson()
     }
 
+    /**
+     Modify an instance (only the fields that are present in the request)
```
Comment (tannernelson) [inline Sources/Development/Controllers/UserController.swift]: Spacing is off here.

File: Sources/Development/Controllers/UserController.swift [hunk 2]
```diff
@@ -53,4 +60,12 @@ class UserController: Controller {
         return user
     }
 
+    /**
+        Delete all instances.
+     */
+    func destroyAll(_ request: Request) throws -> ResponseRepresentable {
```
Comment (tannernelson) [inline Sources/Development/Controllers/UserController.swift]: This will conflict with #296

**Comment by tannernelson** [comment]
It's my understanding that `PUT/PATCH` can be used interchangeably, i.e. they are equivalent.

**Comment by sarbogast** [comment]
Not at all. Put updates an entire resource. It's like replacing one version of an entity by another one, and the resource in the request must have all the required fields. In a patch, you are only supposed to change the fields that are provided in the request, and leave the others alone. So a patch request can be partial. It allows for updates with lower bandwidth consumption. See the difference in implementation here: https://github.com/sarbogast/vapor-todobackend/blob/master/App/Controllers/TodoController.swift#L91

Also PUT is idempotent, but not PATCH. For more information: http://restcookbook.com/HTTP%20Methods/patch/

**Comment by Evertt** [comment]
@tannernelson Nope, I've learned that according to the HTTP spec there's a difference between `PATCH` and `PUT`. The difference is:

The data sent by `PUT` completely replaces all the data for that resource. While the data sent by `PATCH` only replaces that part of the data for that resource.

Let me give you an example. Say I have a `Post` model which has a `title` field and a `body` field. When I send a json `{"body": "This is my new body"}` through a `PATCH` request then the body of that post will be updated to that new body, while the title will be left alone. If, however, I sent that json through a `PUT` request then the `title` will be set to `nil`.

**Comment by tannernelson** [comment]
Beautiful. Looks like Laravel gets those wrong:

```
PUT/PATCH	/photo/{photo}	update	photo.update
```

Let's make sure Vapor follows the HTTP spec!

**Comment by sarbogast** [comment]
Well, as a matter of fact, if your resource is a binary resource like an image, then effectivement PUT and UPDATE do the same thing.

**Comment by LoganWright** [comment]
@tannernelson @sarbogast @Evertt 

After reading the conversation above, I am second guessing the name `update`. It's not clear whether that method should be treated like a replace or not. On one hand, it doesn't really matter because if it's a put and all fields are present, it will update all of the fields, on the other hand, I think it would be more clear if we renamed it to `replace`. 

How do people feel about that, or is it way off base?

**Comment by Evertt** [comment]
@LoganWright I was thinking the same thing. Even though it goes against the conventions of most frameworks, I vote for renaming it to `replace` and having another `update` or `modify` method for the `PATCH` request.

**Comment by Evertt** [comment]
Or how about something like this:

* **replace** for `PUT` requests
* **modify** for `PATCH` requests
* **update** for both of those, like is the convention in almost all other frameworks, for people who really don't want to think / care about the difference between `PATCH` and `PUT`

**Comment by LoganWright** [comment]
@Evertt I was thinking in a similar direction, but keep getting stuck on behavior a little bit. For example, if we receive a PUT request, should we route it to `update` _and_ `replace`? and in what order. Maybe we just do that and don't worry about it, any thoughts on this?

**Comment by Evertt** [comment]
@LoganWright as long as we document the difference between those methods very well and how they should be used, I don't think we need to have any defined behavior for when somebody uses both the `update` method AND one of the others. I think if we say in the documentation that you shouldn't use those method together then it's the responsibility of the developer to not do that.

But just as a thought experiment, just in the rare edge case someone **does** use `update` plus one or both of the others then I assume s/he has the following reason: to keep their code DRY and by putting code in `update` that they would otherwise have to copy in both `replace` and `modify`.

If that's the case then I would say, first route to `update`, and then to the other method. What do you think?

**Comment by LoganWright** [comment]
These commits are tagged on #346 as one PR.

**Pull request closed without merging**